### ISSUE TITLE
buildSrc and buildSrc-like included builds spin up Kotlin daemon that for the most users used only once

### ISSUE THREAD

**Issue by ALikhachev** [gradle/gradle#29331]
### Current Behavior

1. The `org.gradle.kotlin.kotlin-dsl` plugin applies `org.jetbrains.kotlin.jvm` under the hood of some version (let's call it Build Logic Kotlin).
2. The user project applies another version of `org.jetbrains.kotlin.jvm`/`org.jetbrains.kotlin.multiplatform` (let's call it User Kotlin)
3. During the build, both Build Logic Kotlin and User Kotlin may spawn Kotlin daemons to run the compilation.
4. Both daemons will live by default for 2 hours, however, the Kotlin daemon of Build Logic Kotlin most of the time is used once or a couple of times, occupying the system memory.

### Expected Behavior

There are several not mutually exclusive options on how this could be actioned:
1. Lower the lifetime of the Kotlin daemon of Build Logic Kotlin from 2 hours, let's say, to 30 minutes. This can be controlled by a system property, `kotlin.daemon.options=autoshutdownIdleSeconds=1800` on the spawning process level (Gradle daemon in this case).
2. Migrate to in-process compilation for build logic. KGP provides an experimental way of running compilation (that is implemented via KGP's classloader managing logic + workers API with no isolation) that supports in-process incremental-compilation properly. For this to work, the project extra property `kotlin.compiler.runViaBuildToolsApi=true` should be set. KGP may provide a better programmatic way to enable it if needed. As was discussed before, this option may lead to higher memory usage for the Gradle daemon, thus it may be first implemented via a feature flag of the kotlin-dsl plugin and evaluated before using as the default option.

### Context

Current logic of searching a compatible Kotlin daemon:
1. If there's a Kotlin daemon with the same classpath and the same or bigger memory requirement, we can reuse it
2. If there's a Kotlin daemon with the same classpath, but lower memory requirements than ours, we spawn a new daemon, but it automatically takes all the clients of the old daemon so that it can be gently terminated

Currently, Kotlin daemon does not track overall memory pressure or other global characteristics to try minimizing the number of daemons running.

**Comment by eskatos** [comment]
Isn't there any other mechanism to set the IDLE time of the Kotlin compiler daemon other than a jvm-wide system property?

If we wanted to set it only for build-logic related kotlin compiler daemons we'd need a way to set this at the KGP project model level instead. A jvm-wide setting would apply to production code compilation too. Think about included builds for build logic for which execution can be interleaved with production code compilation.

**Comment by eskatos** [comment]
> Currently, Kotlin daemon does not track overall memory pressure or other global characteristics to try minimizing the number of daemons running.

Gradle do that for its daemon and other worker processes. Is there any YT issue we could track about adding this capability to the Kotlin compiler daemon?

**Comment by ALikhachev** [comment]
> > Currently, Kotlin daemon does not track overall memory pressure or other global characteristics to try minimizing the number of daemons running.
> 
> Gradle do that for its daemon and other worker processes. Is there any YT issue we could track about adding this capability to the Kotlin compiler daemon?

Particularly for this one, no. Could you please file a new ticket and provide there some details on how this capability works in Gradle? We would greatly appreciate learning from your experience.

> Isn't there any other mechanism to set the IDLE time of the Kotlin compiler daemon other than a jvm-wide system property?
> 
> If we wanted to set it only for build-logic related kotlin compiler daemons we'd need a way to set this at the KGP project model level instead. A jvm-wide setting would apply to production code compilation too. Think about included builds for build logic for which execution can be interleaved with production code compilation.

Unfortunately, we currently lack such mechanisms. However, we are ready to introduce more granular configuration options for the Kotlin daemon. We are considering exposing this configuration through the same mechanism we use for other KGP properties. This mechanism first checks `project.extra` (which works as well for project-level `gradle.properties`), then `providerFactory.gradleProperty`, and finally `local.properties`. Would this approach work for you?

**Comment by exijn** [comment]
Hi, I'd like to investigate this issue and see whether I can contribute a fix.

I'll first create a minimal reproducer covering both `buildSrc` and an included build used for build logic. I'll record when the Kotlin compiler daemon is started, how long it remains alive, and whether its lifecycle can be configured independently from daemons used for production Kotlin compilation.

Since this may require coordination with the Kotlin Gradle Plugin, I'll report the exact ownership boundary and measurements here before proposing code changes. If a Gradle-side change is appropriate, I'll add focused integration coverage and prepare the smallest possible PR.

Is there a preferred profiling or integration-test setup for investigating Gradle worker and compiler daemon lifecycles?
